controllers/nstemplateset/namespaces.go

```diff
@@ -189,31 +191,50 @@ func (r *namespacesManager) ensureInnerNamespaceResources(ctx context.Context, n
 		return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to process template for namespace '%s'", nsName)
 	}
 
-	if currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]; exists && currentRef != "" && currentRef != tierTemplate.templateRef {
-		logger.Info("checking obsolete namespace resources", "spacename", nsTmplSet.GetName(), "tier", nsTmplSet.Spec.TierName, "type", tierTemplate.typeName)
-		if err := r.setStatusUpdatingIfNotProvisioning(ctx, nsTmplSet); err != nil {
-			return err
-		}
-		currentTierTemplate, err := getTierTemplate(ctx, r.GetHostCluster, currentRef)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to retrieve current TierTemplate with name '%s'", currentRef)
-		}
-		currentObjs, err := currentTierTemplate.process(r.Scheme, map[string]string{
-			SpaceName: nsTmplSet.GetName(),
-		}, template.RetainAllButNamespaces)
-		if err != nil {
-			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, r.setStatusUpdateFailed, err, "failed to process template for TierTemplate with name '%s'", currentRef)
+	// Check if we need to delete any obsolete resources
+	// First check if the namespace already has template reference meaning it's not freshly created and has not got any inner resources applied yet
+	currentRef, exists := namespace.Labels[toolchainv1alpha1.TemplateRefLabelKey]
+	if exists {
+		// Now check if there is any featured object to be deleted in case the feature annotation does not present in the NSTemplateSet anymore
+		toDeleteObsoleteFeaturedObjects := false
+		featureStr := nsTmplSet.Annotations[toolchainv1alpha1.FeatureToggleNameAnnotationKey]
+		features := strings.Split(featureStr, ",")
```

Comment:
this should happen only if `currentRef != tierTemplate.templateRef`, right? Otherwise, it would be triggered everytime when there is some object with a feature not enabled for the given NSTemplateSet

Reply:
If we do it in case the remplate ref was changed only then it wouldn't cover the cases of disabled/enabled feature for existing NSTemplateSets. 
So, you are correct that it triggers every time there is a feature object without the corresponding feature in the NSTemplateSet.

But it doesn't do too much. What it actually triggers:
- another template processing to get the list of objects defined in templates. It's already done multiple times (!) by this point. So we are doing it one more time.
- attempt to delete a feature object even if it doesn't actually exist. Should be fine because of the cache though. We already list all existing Roles and RBs. Every time. So it should not a big deal with the attempt deletion.

Is it perfect/optimal? No, it's not.
Does it make any significant performance impact? Unless I'm missing something, it should not make a noticeable difference.

Reply:
ok, now I understand why you did it - to make sure that we always delete all feature-disabled resources. 

No, it shouldn't make it much worse. Mainly when we know that we won't have any feature-related resources in DevSandbox templates any time soon. I hope that the next implementation of the controller will make the logic simpler and more effective.

Reply:
I deleted all changes for namespaced resources. The PR supports the cluster scoped resource only now.